### PR TITLE
Add Deputy subsystem for low-pop security

### PR DESCRIPTION
**Pull request by blinkdog**
## What Does This PR Do
This PR creates a Deputy subsystem. When the subsystem detects that the population is too low or that the number of people playing security roles is too low, then non-antags are Deputized.

deputy-alert
NOTE: Thanks to McRamon for this amazing icon.

This PR adds a new configuration variable `USE_DEPUTY_SUBSYSTEM` which is disabled by default. When disabled, the subsystem does NOT fire and no crew deputized at any point during the round.

When enabled, two additional configuration values come into play:
* `DEPUTY_BELOW_PLAYERCOUNT` - When the number of active players falls below this number, non-antags are deputized.
* `DEPUTY_BELOW_SEC_PERCENT` - When the percent of players playing security roles falls below this number, non-antags are deputized. (At 10%, a crew of ten will be deputized unless there is at least 1 security role. At 5%, a crew of twenty will be deputized unless there is at least 1 security role. Etc.)

Either condition can trigger deputy status; low population, low security, or both. For example, if the playercount threshold is set for 10 and there are 5 people (2 security), the non-antagonist crew will be deputized due to low population even though there are enough security roles.

## Why It's Good For The Game
During low population and/or when there are no security roles present, antagonists face little to no challenge, as other roles are following valid hunting rules designed for higher populations.

This gives crew a "License to Valid Hunt" because an antagonist being hunted by 5 security and 50 crew is a very different situation than an antagonist being hunted by 0 security and 5 crew.

It keeps both antagonist and crew on their toes, and gives the antagonist an actual challenge during the round.

## Changelog
:cl:
add: Adds Deputy subsystem to create a low-pop vigilante security force
/:cl:

### CODE REVIEW

**Comment by trololiver112** [comment]
From the discord: 


> Looking at that PR... there is no way for other crew to check right?

> Even animals or golems get it

File: code/controllers/subsystem/deputy.dm [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+/datum/controller/subsystem/deputy/fire()
+	// first determine how many active security people we have on station
+	var/list/active_with_role = number_active_with_role()
+	var/num_security = active_with_role["Security"]
+
+	// next determine how many active players we have
+	var/num_players = active_with_role["Any"]
+
+	// if there are no players, there is nobody to deputize
+	if(num_players < 1)
```
Comment (ZomgPonies) [inline code/controllers/subsystem/deputy.dm:26]: if(!num_players)

File: code/controllers/subsystem/deputy.dm [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	var/sec_percent = PERCENT(num_security / num_players)
+	if(sec_percent < config.deputy_below_sec_percent)
+		deputized = TRUE
+
+	// determine if the crew is changing status
+	var/gain_deputy = (!last_deputized && deputized)
+	var/lose_deputy = (last_deputized && !deputized)
+
+	// for every player in the game
+	for(var/mob/M in GLOB.player_list)
```
Comment (ZomgPonies) [inline code/controllers/subsystem/deputy.dm:46]: for(var/mob/living/carbon/human/M in GLOB.player_list)

Deputizing ghosts, Ian and terror spiders can be a problem. Probably would need as seperate checks for borgs if needed.

**Comment by SteelSlayer** [comment]
Love the sheriff badge sprite.

File: code/controllers/subsystem/deputy.dm (hunk 1 above)
Comment (blinkdog) [reply code/controllers/subsystem/deputy.dm:26]: I'll just leave this as-is. I'm not a fan of implicit casts from integer to boolean.

File: code/controllers/subsystem/deputy.dm (hunk 2 above)
Comment (blinkdog) [reply code/controllers/subsystem/deputy.dm]: Hopefully I got the types right this time around.

**Comment by ZomgPonies** [comment]
Alright looks like we'll be ready to testmerge this. Should be easy to test with the lowpop.

**Comment by DoctorDrugs** [comment]
you're most likely going to want to make it add an admin log, a conversion log in Log Viewer, or another method for when someone is been made into a deputy to make sure admins are aware that someone attacking/arresting/dealing with an antag is a deputy. otherwise admins will have to check VV in order to ascertain that info which isn't ideal if it's something that needs to be dealt with in a timely manner as it will slow down investigations/issue resolution to poke around in VV instead of seeing it in the same panel they're already probably looking at.

**Comment by blinkdog** [comment]
Instead of treating Deputy as a per-mob thing (person X is deputized, person Y is not), now it's a game state thing.
All connected clients now see if regular crew are allowed to engage antagonists or not.
I did add some logging to help resolve disputes, but given the global nature of the alert, I don't anticipate many disputes.

**Comment by blinkdog** [comment]
The problem is that the rules are written for a different population level than we have.
We don't need this PR, we need rules appropriate to our population level.

**Pull request closed without merging**